python/pylibcudf/pylibcudf/column.pyx

```diff
@@ -1319,3 +1347,263 @@ def is_c_contiguous(
             return False
         cumulative_stride *= dim
     return True
+
+
+cdef bint _is_valid(const uint8_t* null_bitmap, size_t bit_index):
+    # Example:
+    # - column = [2, None, 3, 1, None, 5, None, 7, 8, None]
+    # - index  = [0, 1,    2, 3, 4,    5, 6,    7, 8, 9]
+    #
+    # Row values:
+    # - row 0 -> valid -> 1
+    # - row 1 -> null  -> 0
+    # - row 2 -> valid -> 1
+    # - row 3 -> valid -> 1
+    # - row 4 -> null  -> 0
+    # - row 5 -> valid -> 1
+    # - row 6 -> null  -> 0
+    # - row 7 -> valid -> 1
+    # - row 8 -> valid -> 1
+    # - row 9 -> null  -> 0
+    #
+    # Build the bitmap:
+    # - index     0,1,2,3,4,5,6,7       76543210 (row_index -> bit_index)
+    # - Byte 0 = [1,0,1,1,0,1,0,1] -> 0b10101101
+    # - Byte 1 = [1,0,0,0,0,0,0,0] -> 0b00000001
+    #
+    # Check the 7th element:
+    # row_index = 6
+    # offset = 0
+    # bit_index = offset + row_index = 6
+    # 1. Choose the byte
+    #    bit_index >> 3 = 0
+    #    null_bitmap[0] = 0b10101101
+    # 2. Get the bit position within that byte (bit_index mod 8)
+    #    bit_index & 7 = 6 & 7 = 6
+    # 3. Shift the target bit to the least significant bit position
+    #    0b10101101 >> 6 = 0b00000010
+    # 4. Remove all other bits
+    #    0b00000010 & 1 = 0 (=> the 7th element is null)
+    if null_bitmap == NULL:
+        return 1
+    return ((null_bitmap[bit_index >> 3] >> (bit_index & 7)) & 1) != 0
+
+cdef void _set_item(list out, Py_ssize_t i, object obj):
+    # SET_ITEM macro "steals" a reference to obj, so we INCREF it first.
+    Py_INCREF(obj)
+    PyList_SET_ITEM(out, i, obj)
+
+cdef list _arrow_to_pylist_bool8(ArrowArray* arr):
+    cdef size_t n = <size_t>arr.length
+    cdef list out = <list>PyList_New(<Py_ssize_t>n)
+    if out is None:
+        raise MemoryError("Unable to create new python List")
+
+    cdef const void** buffers = arr.buffers
+    cdef int64_t offset = arr.offset
+    cdef const uint8_t* null_bm = NULL
+    cdef const uint8_t* data_bm = NULL
+    cdef size_t i, bit_index
+    cdef object val
+
+    if n == 0:
+        return out
+
+    if arr.n_buffers >= 1 and buffers[0] != NULL and arr.null_count != 0:
+        null_bm = <const uint8_t*>buffers[0]
+
+    if arr.n_buffers < 2 or buffers[1] == NULL:
+        for i in range(n):
+            _set_item(out, <Py_ssize_t>i, None)
+        return out
+
+    data_bm = <const uint8_t*>buffers[1]
+
+    for i in range(n):
+        bit_index = <size_t>(offset + <int64_t>i)
+        if not _is_valid(null_bm, bit_index):
+            _set_item(out, <Py_ssize_t>i, None)
+            continue
+
+        val = ((data_bm[bit_index >> 3] >> (bit_index & 7)) & 1) != 0
+        _set_item(out, <Py_ssize_t>i, val)
+
+    return out
+
+
+cdef list _arrow_to_pylist_string(ArrowArray* arr):
```

Comment:
This function does have some more differences from the numeric/bool cases, but I still think it's similar enough to be worth combining them. You'll need to maintain a few extra variables for every invocation on non-numeric data (like the offsets) and those will just be null for those types, but the rest of the logic can just be shoved into an else clause for string data.